File: software/matlab/plot_results.m
function plot_results(SER, SNR_dB, fig_select, p)
%PLOT_RESULTS  Generate SER vs SNR plots matching paper figures.
%  Compatible with MATLAB R2014a and later.

figure('Name', fig_select, 'Position', [100 100 680 520]);
hold on;
grid on;

markers = {'o-','s--','d:','v-.','^-','<--','>:','p-.','h-','*--'};
colors  = lines(12);

alg_names  = fieldnames(SER);
n_algs     = numel(alg_names);
leg_labels = cell(1, n_algs);   % collect labels in the same order as plots

for a = 1:n_algs
    tag = alg_names{a};
    ser = SER.(tag);
    ser(ser == 0) = NaN;
    semilogy(SNR_dB, ser, markers{mod(a-1, numel(markers))+1}, ...
             'Color', colors(a,:), 'LineWidth', 1.5, 'MarkerSize', 7);
    leg_labels{a} = label_name(tag);
end

% Pass label strings explicitly — reliable in all MATLAB versions
legend(leg_labels, 'Location', 'SouthWest', 'FontSize', 9);

xlabel('SNR [dB]', 'FontSize', 12);
ylabel('SER',      'FontSize', 12);
xlim([SNR_dB(1) SNR_dB(end)]);
ylim([1e-4 1]);
set(gca, 'YScale', 'log', 'XGrid', 'on', 'YGrid', 'on');

title_str = sprintf('SER vs SNR  |  %s  M=%d C=%d K=%d IoT=%ddB %s', ...
            fig_select, p.M, p.C, p.K, p.IoT_dB, p.mod);
title(title_str, 'FontSize', 10);

out_name = sprintf('SER_%s_M%d_C%d_K%d.png', fig_select, p.M, p.C, p.K);
print(gcf, out_name, '-dpng', '-r150');
fprintf('Figure saved: %s\n', out_name);
end


function str = label_name(tag)
%LABEL_NAME  Human-readable legend labels.
map = { ...
    'LMMSE',     'LMMSE (centralised)'; ...
    'ZF',        'ZF (centralised)'; ...
    'MMSE_AWGN', 'MMSE(AWGN)'; ...
    'BDAC',      'BDAC-MMSE'; ...
    'BCD1',      'BCD-MMSE: Iter 1'; ...
    'BCD2',      'BCD-MMSE: Iter 2'; ...
    'BCD3',      'BCD-MMSE: Iter 3'; ...
    'BCD4',      'BCD-MMSE: Iter 4'; ...
    'LRD1',      'BCD-MMSE(LRD): Iter 1'; ...
    'LRD2',      'BCD-MMSE(LRD): Iter 2'; ...
    'LRD3',      'BCD-MMSE(LRD): Iter 3'; ...
    'LRD4',      'BCD-MMSE(LRD): Iter 4'; ...
    'sDR',       'sDR-MMSE'; ...
    'cDR',       'cDR-MMSE'; ...
};
idx = strcmp(map(:,1), tag);
if any(idx)
    str = map{idx, 2};
else
    str = tag;
end
end

File: software/python/python/plot_results.py
"""Plot SER vs SNR results matching paper style."""

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt


# Style definitions matching paper figures
STYLES = {
    'LMMSE':    dict(color='blue',    marker='o', ls='-',  lw=2, label='LMMSE (centralised)'),
    'ZF':       dict(color='#228B22', marker='s', ls='--', lw=2, label='ZF (centralised)'),
    'MMSE_AWGN':dict(color='gray',    marker='^', ls='-.', lw=1.5, label='MMSE (AWGN)'),
    'BDAC':     dict(color='brown',   marker='v', ls=':',  lw=1.5, label='BDAC-MMSE'),
    'BCD1':     dict(color='red',     marker='d', ls=':',  lw=1.5, label='BCD-MMSE: Iter 1'),
    'BCD2':     dict(color='#FF6600', marker='p', ls=':',  lw=1.5, label='BCD-MMSE: Iter 2'),
    'BCD3':     dict(color='#CC00CC', marker='h', ls=':',  lw=1.5, label='BCD-MMSE: Iter 3'),
    'BCD4':     dict(color='#8B0000', marker='*', ls=':',  lw=1.5, label='BCD-MMSE: Iter 4'),
    'LRD1':     dict(color='#00BFFF', marker='d', ls='-.', lw=1.5, label='BCD-MMSE(LRD): Iter 1'),
    'LRD2':     dict(color='#1E90FF', marker='p', ls='-.', lw=1.5, label='BCD-MMSE(LRD): Iter 2'),
    'LRD3':     dict(color='#4169E1', marker='h', ls='-.', lw=1.5, label='BCD-MMSE(LRD): Iter 3'),
    'LRD4':     dict(color='cyan',    marker='v', ls='-.', lw=1.5, label='BCD-MMSE(LRD): Iter 4'),
    'sDR':      dict(color='magenta', marker='^', ls='-',  lw=1.5, label='sDR-MMSE'),
    'cDR':      dict(color='#FFD700', marker='<', ls='--', lw=1.5, label='cDR-MMSE'),
}


def plot_results(SER: dict, SNR_dB: np.ndarray, fig_select: str,
                 cfg=None, save_path: str = None):
    """Generate SER vs SNR plot.

    Parameters
    ----------
    SER : dict mapping algorithm name → np.ndarray of SER values
    SNR_dB : np.ndarray
    fig_select : str
    cfg : SimConfig (optional, for title info)
    save_path : str (optional, defaults to SER_{fig_select}.png)
    """
    fig, ax = plt.subplots(figsize=(10, 7))

    for alg_name, ser_vals in SER.items():
        style = STYLES.get(alg_name, dict(marker='x', ls='-', lw=1, label=alg_name))
        # Clip zeros for log scale
        ser_plot = np.where(ser_vals > 0, ser_vals, np.nan)
        ax.semilogy(SNR_dB, ser_plot, markersize=7, **style)

    ax.set_xlabel('SNR [dB]', fontsize=13)
    ax.set_ylabel('SER', fontsize=13)

    if cfg is not None:
        title = (f'SER vs SNR | {fig_select}  '
                 f'M={cfg.M} C={cfg.C} K={cfg.K} IoT={cfg.IoT_dB}dB {cfg.mod}')
    else:
        title = f'SER vs SNR | {fig_select}'
    ax.set_title(title, fontsize=14)

    ax.legend(loc='lower left', fontsize=9, framealpha=0.9)
    ax.grid(True, which='both', alpha=0.3)
    ax.set_xlim(SNR_dB[0], SNR_dB[-1])
    ax.set_ylim(1e-4, 1)

    plt.tight_layout()

    if save_path is None:
        save_path = f'SER_{fig_select}.png'
    fig.savefig(save_path, dpi=150)
    print(f"Plot saved to {save_path}")
    plt.close(fig)
